lib/rules/no-template-children.js

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+'use strict';
+
+const { docUrl } = require('../util/doc-url');
+
+module.exports = {
+    meta: {
+        docs: {
+            description: 'prevent accessing the immediate children of this.template',
+            category: 'LWC',
+            recommended: true,
+            url: docUrl('no-template-children'),
+        },
+
+        schema: [],
+    },
+
+    create(context) {
+        const disallowedProperties = ['firstElementChild', 'firstChild', 'children', 'childNodes'];
+
+        /**
+         * Returns true if this is a `this.template` expression
+         */
+        function isThisDotTemplateExpression(node) {
+            return (
+                node.type === 'MemberExpression' &&
+                node.object.type === 'ThisExpression' &&
+                node.property.type === 'Identifier' &&
+                node.property.name === 'template'
+            );
+        }
+
+        /**
+         * Returns true if this is an ObjectExpression property that should be disallowed for `this.template`
+         * E.g.: `const { firstChild } = this.template`
+         */
+        function propertyIsDisallowedForThisDotTemplate(prop) {
+            return prop.key.type === 'Identifier' && disallowedProperties.includes(prop.key.name);
+        }
+
+        /**
+         * Returns true if this is an ObjectExpression property that should be disallowed for `this`
+         * E.g.: `const { template: { firstChild } } = this`
+         */
+        function propertyIsDisallowedForThis(prop) {
+            const { key, value } = prop;
+            return (
+                key.type === 'Identifier' &&
+                key.name === 'template' &&
+                value.type === 'ObjectPattern' &&
+                value.properties.some(propertyIsDisallowedForThisDotTemplate)
+            );
+        }
+
+        return {
+            MemberExpression(node) {
```

Comment:
Would it be possible to only use the `ThisExpression` visitor instead of `MemberExpression` and `VariableDeclarator` to do these validations?

Reply:
I suppose, but then I would need to check the parent of the `ThisExpression`, right? I guess I assumed that ESLint plugins should avoid accessing the parent (no idea why I thought that though!).

Reply:
Fixed it so that it's using `ThisExpression`. Also added some more tests.